### ISSUE TITLE
Pessimistic timeout not behaving as expected on truly asynchronous delegates [if onTimeoutAsync delegate awaits the timed-out task]

### ISSUE THREAD

**Issue by utkarsh5k** [App-vNext/Polly#602]
**Summary**: 

I don't feel like this is a duplicate of [this issue](https://github.com/App-vNext/Polly/issues/340). I am trying to use pessimistic timeouts on my async calls, but it seems like everything including the timeout is still proceeding in a synchronous fashion, and that my flow waits for the entire duration before resuming execution. 

What I basically have (code attached) is a for loop that executes a set of delay tasks asynchronously. The timeout is set to 2 seconds and I decide the time to wait for randomly. I expected that if waiting time for a task is more than 2 seconds, it would throw in 2 seconds and I would resume with my other tasks. However, from the output (attached), it seems like it is executing in a completely synchronous fashion. 

Is my understanding incorrect regarding the pessimistic timeout? Or if there is some mistake in the way I am using it, please point it out. 

Thanks. 

**Expected behavior**: 
If my understanding is correct, if one task is supposed to wait for 10 seconds and the timeout is configured to 2 seconds, the next task should start at +2 instead of +10 seconds.

**Actual behaviour**:
Console output for a sample run. It seems like each task is waiting for its previous task to finish, post which an exception for the timeout is thrown, and only then the execution seems to move forward with the remaining tasks. 

```
Going to begin task with delay 6000 at 2/26/2019 4:15:48 PM and key 0
Done task 2/26/2019 4:15:54 PM with delay 6000 and key 0
Completed  2/26/2019 4:15:54 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 0 2/26/2019 4:15:54 PM
Going to begin task with delay 4000 at 2/26/2019 4:15:54 PM and key 1
Done task 2/26/2019 4:15:58 PM with delay 4000 and key 1
Completed  2/26/2019 4:15:58 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 1 2/26/2019 4:15:58 PM
Going to begin task with delay 1000 at 2/26/2019 4:15:58 PM and key 2
Done task 2/26/2019 4:15:59 PM with delay 1000 and key 2
Going to begin task with delay 1000 at 2/26/2019 4:15:59 PM and key 3
Done task 2/26/2019 4:16:00 PM with delay 1000 and key 3
Going to begin task with delay 4000 at 2/26/2019 4:16:00 PM and key 4
Done task 2/26/2019 4:16:04 PM with delay 4000 and key 4
Completed  2/26/2019 4:16:04 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 4 2/26/2019 4:16:04 PM
Going to begin task with delay 6000 at 2/26/2019 4:16:04 PM and key 5
Done task 2/26/2019 4:16:10 PM with delay 6000 and key 5
Completed  2/26/2019 4:16:10 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 5 2/26/2019 4:16:11 PM
Going to begin task with delay 1000 at 2/26/2019 4:16:11 PM and key 6
Done task 2/26/2019 4:16:12 PM with delay 1000 and key 6
Going to begin task with delay 2000 at 2/26/2019 4:16:12 PM and key 7
Done task 2/26/2019 4:16:14 PM with delay 2000 and key 7
Completed  2/26/2019 4:16:14 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 7 2/26/2019 4:16:14 PM
Going to begin task with delay 3000 at 2/26/2019 4:16:14 PM and key 8
Done task 2/26/2019 4:16:17 PM with delay 3000 and key 8
Completed  2/26/2019 4:16:17 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 8 2/26/2019 4:16:17 PM
Going to begin task with delay 5000 at 2/26/2019 4:16:17 PM and key 9
Done task 2/26/2019 4:16:22 PM with delay 5000 and key 9
Completed  2/26/2019 4:16:22 PM
Polly.Timeout.TimeoutRejectedException: The delegate executed asynchronously through TimeoutPolicy did not complete within the timeout. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 45
   --- End of inner exception stack trace ---
   at Polly.Timeout.AsyncTimeoutEngine.<ImplementationAsync>d__0`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Timeout\AsyncTimeoutEngine.cs:line 53
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Polly.AsyncPolicy.<ExecuteAsync>d__12.MoveNext() in C:\projects\polly\src\Polly.Shared\AsyncPolicy.ExecuteOverloads.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at TimeoutPolicy.Program.<StartAction>d__2.MoveNext() in D:\Misc\TimeoutPolicy\TimeoutPolicy\Program.cs:line 37 ITERATION: 9 2/26/2019 4:16:22 PM
Came here 2/26/2019 4:16:22 PM
Action done 2/26/2019 4:16:22 PM
```

**Steps / Code to reproduce the problem**: 
```
namespace TimeoutPolicy
{
    class Program
    {
        static Random rand = new Random();
        static void Main(string[] args)
        {
            try
            {
                StartAction().Wait();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString() + " " + DateTime.Now);
            }
            Console.WriteLine("Action done" + DateTime.Now);
            Console.ReadKey();
        }
        private static async Task StartAction()
        {
            var policy = Policy.TimeoutAsync(2, Polly.Timeout.TimeoutStrategy.Pessimistic, OnTimeoutAsync);

            for (int i = 0; i < 10; i++)
            {
                try
                {
                    Context context1 = new Context(i.ToString());
                    await policy.ExecuteAsync(async (context) => await DoSomething(context1), new Dictionary<string, object>());
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.ToString() + " ITERATION: " + i + " " + DateTime.Now);
                }
            }

            Console.WriteLine("Came here " + DateTime.Now);
        }
        private static async Task DoSomething(Context context)
        {
            string log = "";
            int delay = rand.Next(1, 11) * 1000;
            log = string.Format("Going to begin task with delay {0} at {1} and key {2}", delay, DateTime.Now, context.OperationKey);
            Console.WriteLine(log);
            await Task.Delay(delay);
            log = string.Format("Done task {0} with delay {1} and key {2}", DateTime.Now, delay, context.OperationKey);
            Console.WriteLine(log);
        }

        public static async Task OnTimeoutAsync(Context context, TimeSpan timeSpan, Task task)
        {
            await task.ContinueWith(abandonedTask =>
            {
                if (abandonedTask.IsFaulted)
                {
                    Console.WriteLine("Abandoned {0} {1}", context.OperationKey, DateTime.Now);
                }
                else if (abandonedTask.IsCanceled)
                {
                    // This is potentially unreachable code
                    Console.WriteLine("Canceled {0} {1}", context.OperationKey, DateTime.Now);
                }
                else if (abandonedTask.IsCompleted)
                {
                    Console.WriteLine("Completed {0} {1}", context.OperationKey, DateTime.Now);
                }
            });
        }
    }
}
```

**Comment by reisenberger** [comment]
Thanks @utkarsh5k for the exemplary repro code and accompanying notes. I believe the behaviour you are seeing is because the `OnTimeoutAsync(...)` method is waiting on the completion of the timed-out task (the original delegate executed through the policy), by doing `await task.ContinueWith(abandonedTask =>`.  **TL;DR** Delete `await` before `task.ContinueWith(...)` and it should behave as expected.

---

### Demonstration

(1) Try adding a statement `Console.WriteLine("Timed out {0} {1}", context.OperationKey, DateTime.Now);` as the first line within `OnTimeoutAsync(...)`, before `await task.ContinueWith(...)`. Keep `await task.ContinueWith(...)` for now. The output should then show that the timeout is occurring at the expected time.  (Overall, the operation will still complete with the delays posted in the question because the `OnTimeoutAsync(...)` still contains `await task.ContinueWith(...)`.)

(2) Delete the whole `await task.ContinueWith(...)` statement but keep the statement added above. The output should show the timeouts occurring correctly, and the overall execution should finish in a timely manner, as the walked-away-from executions are no longer awaited (but we've lost the abandoned/cancelled/completed logging and our guard against `UnobservedTaskException`s).

(3) Reinstate the `task.ContinueWith(...)` statement but don't await it (omit `await`).  The `OnTimeoutAsync(...)` will attach task continuations (in TPL-style) to the walked-away-from `Task`s, but not await them.  The overall execution should finish in a timely manner.  The `task.ContinueWith(...)` continuations will execute whenever the walked-away-from tasks (later) complete, so the console output from them should appear interspersed in the output sequence at the times those walked-away-from Tasks eventually complete.  This should demonstrate what we mean by pessimistic timeout _walking away from_ (but not being able to cancel) the uncancellable delegates; and how the task continuation later 'mops up' their completion.  All the continuations from timed-out/walked-away-from tasks should eventually show in the console output, provided you don't let `Program.Main(...)` terminate first.

---

Let me know if that ^ works as described. To see why the original behaviour was happening, see the [source code here](https://github.com/App-vNext/Polly/blob/df33c17904e4efcdbfd66caeb9be862affcf3091/src/Polly.Shared/Timeout/AsyncTimeoutEngine.cs#L50-L54).  On timeout, the policy [`await`s your `OnTimeoutAsync(...)` delegate](https://github.com/App-vNext/Polly/blob/df33c17904e4efcdbfd66caeb9be862affcf3091/src/Polly.Shared/Timeout/AsyncTimeoutEngine.cs#L52) before throwing to exit indicating the timeout.  Since the original `OnTimeoutAsync(...)` code was doing `await task.ContinueWith(...)` where `task` is the blocking `DoSomething(...)`, the execution was naturally blocked the full 6-seconds, 4-seconds, whatever ... before continuing to the next iteration of the loop.

**Comment by utkarsh5k** [comment]
Works as expected! Thanks for the explanation, @reisenberger!
